### ISSUE TITLE
BO - Stocks - product list content is not loading or delayed long time

### ISSUE THREAD

**Issue by prestaalba** [PrestaShop/PrestaShop#24975]
#### Describe the bug

In backoffice Stocks manager panel, product list content is not loading or delayed long time.

#### Expected behavior

Product list content on Stocks manager panel should load on a reasonable time.

#### Steps to Reproduce

1. Go to menu Catalog -> Stocks.
2. Page loads and product list content starting to load.
3. Content never loads or is delayed long time.

**Screenshots**

Details of ajax request related to this issue:
![1](https://user-images.githubusercontent.com/18466848/122104962-dc27e280-cddd-11eb-848a-c3cf8ad31e20.png)

![2](https://user-images.githubusercontent.com/18466848/122104969-de8a3c80-cddd-11eb-8fb7-81e852f4d92c.png)

![3](https://user-images.githubusercontent.com/18466848/122104981-e0ec9680-cddd-11eb-972d-50e6954e7341.png)

#### Additional information

This is cause because of a bulk update of the field reserved_quantity of all records in table stock_available (https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/StockManager.php#L134) that must search also on orders, order_detail and order_state tables to complete this operation. The StockManager::updateReservedProductQuantity method should not be called without $idProduct (https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Entity/Repository/StockRepository.php#L197) to avoid this unnecesary bulk update. This issue will increase with the time due to new orders and with bigger catalogs make impossible to load this screen.

![4](https://user-images.githubusercontent.com/18466848/122105043-f19d0c80-cddd-11eb-95f2-3004e8564fa3.png)

![5](https://user-images.githubusercontent.com/18466848/122105056-f3ff6680-cddd-11eb-9081-204ad64bc396.png)

* Shop catalog size: 741 products with combinations
* Multi-shop: enabled with 4 shops
* Total orders: 21202
* PrestaShop version: develop
* PHP version: N/A

**Comment by hibatallahAouadni** [comment]
Hello @prestaalba 

This issue has been reported before #24517 but I couldn't reproduce it.
Could you please check the issue and the proposed solution by @alejandrogarciahenao 
Please check and feedback.

Thanks!

**Comment by prestaalba** [comment]
Hello @hibatallahAouadni 

In fact, this is the same issue that https://github.com/PrestaShop/PrestaShop/issues/24517. To reproduce it you must have a shop with medium-high catalog and orders size (as described on this issue Additional information).

The solution proposed by @alejandrogarciahenao make that product list data loads fast, but only because the bulk update process on updateReservedProductQuantity method is completely skipped. Must exist a reason because this method is called, so it shouldn't be skipped.

This should be the correct solution: Update only the products that will be displayed on each page.
https://github.com/PrestaShop/PrestaShop/pull/24989

Regards

**Comment by khouloudbelguith** [comment]
Hi @prestaalba,

Could you please enable profiling by editing the file `/PrestaShop/config/defines.inc.php`
```php
if (!defined('_PS_DEBUG_PROFILING_')) {
    define('_PS_DEBUG_PROFILING_', true);
```
To check

- Load Time 
- Querying Time
- Queries


Thank you!

**Comment by prestaalba** [comment]
Hi @khouloudbelguith :

This investigation was made with profiling enabled, however no clue could be found there, maybe because this issue is related to an update query with subqueries. Anyway, this is a known bug already reported previously. I recommend you to check the pull request to see the possible solution.

Regards.

**Comment by khouloudbelguith** [comment]
Hi @prestaalba,

I have a shop PS1775 with:
- 1368 products
- 81 orders
- 25 shops

The Stocks page => ok
I attached a screen record
https://drive.google.com/file/d/1zwlpKaj7v5MJE3IFsdUfzwL0g6lf96-Q/view?usp=sharing

Thanks to check and feedback

**Comment by prestaalba** [comment]
Hi @khouloudbelguith, 

As you can see on https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/StockManager.php#L134 the update query of this function is directly related to orders and order_detail tables. So, as you have more order records, the longer it will take to execute the query. 81 orders is a tiny amount of records to be able to see the problem. However 20k orders is a common quantity for any shop with maybe 5 years old or less.
This issue is a fact obtained from a real shop and confirmed with a different shop with a similar number of records. Of course, I was never able to reproduce this problem in a store with fewer records.

Anyway, a solution has already been proposed, with no impact on the query, just in the way the function is called. Could you please check it?

Regards.

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/prestashop-core-developers could you please check this performance issue :pray: 
PS: similar to #24517

**Comment by dorkster100** [comment]
I believe the `StockRepository::getData` method should not be updating whole store every time. This is not sustainable, what about moving that logic to the point at which orders are made and manipulated with, maybe Observer pattern? Doing this product by product seems a better solution. Or perhaps we do not need to store those values in db at all and can calculate them each time we are looking at the stock table.

**Comment by kpodemski** [comment]
Yes, I can confirm that there is an issue. It seems that the StockRepository is attempting to regenerate the data for reserved products every time, which is causing a significant decrease in performance and is highly inefficient.

**Comment by jf-viguier** [comment]
I've faced this bug too. @kpodemski do you have any idea on how to solve this ?

**Comment by kpodemski** [comment]
@jf-viguier https://github.com/PrestaShop/PrestaShop/pull/34863 this PR improves situation by a lot but it also requires some more work to make sure it covers all the scenarios needed

**Comment by boo-code** [comment]
Still reproducible on 9.2.x, same call path: `StockRepository::getData()` calls `updatePhysicalProductQuantity()` with no product and no order, so `updateReservedProductQuantity()` runs one correlated sub-select over `orders`, `order_detail` and `order_state` for every `stock_available` row of the shop - on every load and every filter or paging request of the grid.

Measured on a synthetic fixture of 40,000 `stock_available` rows and 20,000 orders, running that exact statement. Both variants warm, after a discarded warm-up pass:

| order lines | whole catalogue (what the grid does) | one page of 50 products |
| --- | --- | --- |
| 40,000 | 316 ms | 9.0 ms |
| 400,000 | 1,238 ms | 4.9 ms |

The first, cold run on the 40,000-line fixture had not finished when it was killed at 157 s. That figure is cold-cache and not the one to quote, but it is the state a merchant meets on a page not recently opened, and it matches "content never loads". Worth noting for the two "cannot reproduce" verdicts on this: #24517 was closed as "most likely a server problem", and the QA check here used a shop with 81 orders - at that volume there is nothing to see.

Something the thread has not established, and it decides whether scoping is safe: `updatePhysicalProductQuantity()` has six callers and **every one of them is already scoped**.

| Caller | Scope |
| --- | --- |
| `classes/PaymentModule.php:793` | `$idOrder` |
| `classes/order/OrderHistory.php:326` | `$idOrder` |
| `src/Adapter/Order/CommandHandler/AbstractOrderCommandHandler.php:57` | `$idOrder` |
| `src/Adapter/Order/OrderProductQuantityUpdater.php:388` | `$idOrder` |
| `src/Adapter/Product/Stock/Repository/StockAvailableRepository.php:316` | `StockId` |
| `src/PrestaShopBundle/Entity/Repository/StockRepository.php:134` | `$idProduct` |

Order creation, status changes, order product edits and product or combination stock edits all keep the values correct incrementally. The grid's pass is a repair pass, not the source of truth - which is the answer to @prestaalba's "must exist a reason because this method is called".

Prepared as a patch on `9.2.x`: scope the pass to the products the page is about to display, and read the refreshed quantities back into the rows already fetched. One grid query, N indexed single-product updates, one small indexed select. @prestaalba's #24989 had the same idea and was closed by @PierreRambaud with "feel free to reopen if you can provide another way" - the difference is that it called `parent::getData()` twice, so the heavy grid query ran on every page load in addition to the syncs.

The one real objection, stated up front: `orderBy()` maps `{physical_quantity}` to `sa.physical_quantity`, a column this pass writes, so sorting the grid by physical quantity now orders on stored values rather than on values recomputed a moment earlier. `{available_quantity}` is `sa.quantity`, which the pass never writes, so that sort is unaffected. Given the table above, a discrepancy needs a write that bypassed all six paths.

Nothing here touches `src/Adapter/StockManager.php`, so there is no overlap with #34863.
